kmath-geometry/src/commonMain/kotlin/scientifik/kmath/geometry/Euclidean3DSpace.kt

```diff
@@ -28,30 +30,29 @@ interface Vector3D : Point<Double>, Vector, SpaceElement<Vector3D, Vector3D, Euc
     override fun Vector3D.wrap(): Vector3D = this
 }
 
-@Suppress("FunctionName")
-fun Vector3D(x: Double, y: Double, z: Double): Vector3D = Vector3DImpl(x, y, z)
-
-val Vector3D.r: Double get() = Euclidean3DSpace.run { sqrt(norm()) }
-
-private data class Vector3DImpl(
-    override val x: Double,
-    override val y: Double,
-    override val z: Double
-) : Vector3D
-
-object Euclidean3DSpace : GeometrySpace<Vector3D> {
+object Euclidean3DSpace : InnerProductSpace<Double, D3, Vector3D> {
```

Comment:
Not sure if this fits inside the scope for this PR, but it would be great if KMath were able to add support for quaternions, possibly by extending [`ComplexNDField.kt`](https://github.com/mipt-npm/kmath/blob/master/kmath-core/src/commonMain/kotlin/scientifik/kmath/structures/ComplexNDField.kt). These are widely used in robotics and computer vision for implementing spatial rotations. You might want to check out [amodeus-science/tensor](https://github.com/amodeus-science/tensor/blob/ebeaaf9ec83ab6be2798a1548f131eeda9a37cd7/src/main/java/ch/ethz/idsc/tensor/qty/QuaternionImpl.java) for a good reference implementation.

Reply:
Quaternions are quite easy compared to N-dimensional geometry. I've created #155 to track that. I do not think they should be connected to complex numbers, but we can use the same boxing-avoiding logic there.

Reply:
Quaternions are implemented as part of `complex` module.